compiler/luci/pass/src/FuseGRUPass.cpp

```diff
@@ -0,0 +1,477 @@
+{
+  assert(graph);
+
+  auto weight_ih_cloned = clone_circleconst(_p->_weight_ih, graph);
+  luci::copy_common_attributes(_p->_weight_ih, weight_ih_cloned);
+
+  auto weight_hh_cloned = clone_circleconst(_p->_weight_hh, graph);
+  luci::copy_common_attributes(_p->_weight_hh, weight_hh_cloned);
+
+  luci::CircleNode *bias_ih_cloned = nullptr;
+  if (_p->_bias_ih != nullptr)
+  {
+    bias_ih_cloned = clone_circleconst(_p->_bias_ih, graph);
+    luci::copy_common_attributes(_p->_bias_ih, bias_ih_cloned);
+  }
+  else
+  {
+    bias_ih_cloned = _p->_pattern_last_node->graph()->nodes()->create<luci::CircleOutputExclude>();
+  }
+
+  luci::CircleNode *bias_hh_cloned = nullptr;
+  if (_p->_bias_hh != nullptr)
+  {
+    bias_hh_cloned = clone_circleconst(_p->_bias_hh, graph);
+    luci::copy_common_attributes(_p->_bias_hh, bias_hh_cloned);
+  }
+  else
+  {
+    bias_hh_cloned = _p->_pattern_last_node->graph()->nodes()->create<luci::CircleOutputExclude>();
+  }
+
+  auto hidden_input_cloned = clone_circleconst(_p->_hidden_input, graph);
+  luci::copy_common_attributes(_p->_hidden_input, hidden_input_cloned);
+
+  // Create and configure new CircleGRU
+  auto circle_gru = _p->_while_node->graph()->nodes()->create<luci::CircleGRU>();
+
+  circle_gru->input(_p->_ifm);
+  circle_gru->hidden_input(weight_ih_cloned);
+  circle_gru->hidden_input_bias(bias_ih_cloned);
+  circle_gru->hidden_hidden(weight_hh_cloned);
+  circle_gru->hidden_hidden_bias(bias_hh_cloned);
+  circle_gru->state(hidden_input_cloned);
+
+  circle_gru->fusedActivationFunction(_p->_act_function);
+  circle_gru->returnSequences(_p->_return_sequences);
+  circle_gru->timeMajor(_p->_time_major);
+
+  circle_gru->name("CircleGRU");
+
+  circle_gru->shape_status(luci::ShapeStatus::UNDEFINED);
+
+  circle_gru->dtype(loco::DataType::FLOAT32);
+
+  return circle_gru;
+}
+
+void FuseGRU::apply()
+{
+  auto graph = _p->_pattern_last_node->graph();
+
+  auto circle_gru = create_circle_gru(graph);
+
+  // set origin
+  std::vector<std::shared_ptr<luci::CircleNodeOrigin>> origin_vec{
+    luci::get_origin(_p->_while_node), luci::get_origin(_p->_while_out_node),
+    luci::get_origin(_p->_weight_hh), luci::get_origin(_p->_weight_ih)};
+
+  luci::add_origin(circle_gru, luci::composite_origin(origin_vec));
+
+  replace(_p->_pattern_last_node).with(circle_gru);
+}
+
+} // namespace
+
+namespace
+{
+
+bool fuse_gru(luci::CircleWhileOut *while_out_node)
+{
+  assert(while_out_node);
+
+  // check first pattern
+  GRUPattern1 pattern(while_out_node);
+  if (pattern.matched())
+  {
+    FuseGRU fuse(&pattern);
+    fuse.apply();
+    return true;
+  }
+
+  return false;
+}
+
+} // namespace
+
+namespace luci
+{
+
+bool FuseGRUPass::run(loco::Graph *g)
+{
+  bool changed = false;
+
+  for (auto node : loco::active_nodes(loco::output_nodes(g)))
+  {
+    auto while_out_node = dynamic_cast<luci::CircleWhileOut *>(node);
+    if (not while_out_node)
+      continue;
+
+    if (fuse_gru(while_out_node))
```

Comment:
Q) are other sub graphs like the body graph removed?
I don't recall we had some method to remove graph in the Module.

Reply:
> Q) are other sub graphs like the body graph removed?

Actually - no. We need to support some methods to remove subgraphs in the Module

Reply:
What do you think about add a new option that will leave only the main sub-graph, and delete others? Smth like `--use_only_main_graph` and use fuse_gru and use_only_main_graph together. Otherwise, we need to transfer/save the luci::Module entity to CircleOptimizer to remove unnecessary graphs

Reply:
> Otherwise, we need to transfer/save the luci::Module entity to CircleOptimizer to remove unnecessary graphs

Maybe some kind of eliminate dead sub-graph...
1/ As of now, all sub-graphs having ID > 0 are used by some other sub-graph (usually ID == 0).
(I'm not sure this will hold in the future)
2/ We can use `void CircleOptimizer::optimize(luci::Module *m) const`
3/ Add something like `EliminateDeadSubgraphPass` (maybe with `--elminate-dead-graph` option or run this for default without any options)
4/ Traverse main graph and mark sub-graphs that are used
5/ Delete sub-graphs that are not marked

Can this be feasible ?

Reply:
> Can this be feasible ?

I think yes, good idea to add `EliminateDeadSubgraphPass` and run this for default